File: src/test/java/co/onmind/microhex/infrastructure/controllers/RoleControllerTest.java
package co.onmind.microhex.infrastructure.controllers;

import co.onmind.microhex.application.dto.CreateRoleRequest;
import co.onmind.microhex.application.dto.RoleResponse;
import co.onmind.microhex.application.dto.UpdateRoleRequest;
import co.onmind.microhex.application.handlers.RoleHandler;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

/**
 * Unit tests for RoleController.
 * 
 * @author OnMind (Cesar Andres Arcila Buitrago)
 * @version 2.0.0
 */
@ExtendWith(MockitoExtension.class)
@DisplayName("Role Controller Tests")
class RoleControllerTest {
    
    @Mock
    private RoleHandler roleHandler;
    
    @InjectMocks
    private RoleController roleController;
    
    @Nested
    @DisplayName("Create Role Tests")
    class CreateRoleTests {
        
        @Test
        @DisplayName("Should create role successfully")
        void shouldCreateRoleSuccessfully() {
            // Given
            CreateRoleRequest request = new CreateRoleRequest("ADMIN");
            RoleResponse response = new RoleResponse(1L, "ADMIN", LocalDateTime.now());
            
            when(roleHandler.createRole(any(CreateRoleRequest.class)))
                .thenReturn(ResponseEntity.status(HttpStatus.CREATED).body(response));
            
            // When
            ResponseEntity<RoleResponse> result = roleController.createRole(request);
            
            // Then
            assertEquals(HttpStatus.CREATED, result.getStatusCode());
            assertNotNull(result.getBody());
            assertEquals(1L, result.getBody().getId());
            assertEquals("ADMIN", result.getBody().getName());
        }
        
        @Test
        @DisplayName("Should return conflict when role already exists")
        void shouldReturnConflictWhenRoleAlreadyExists() {
            // Given
            CreateRoleRequest request = new CreateRoleRequest("ADMIN");
            
            when(roleHandler.createRole(any(CreateRoleRequest.class)))
                .thenReturn(ResponseEntity.status(HttpStatus.CONFLICT).build());
            
            // When
            ResponseEntity<RoleResponse> result = roleController.createRole(request);
            
            // Then
            assertEquals(HttpStatus.CONFLICT, result.getStatusCode());
            assertNull(result.getBody());
        }
    }
    
    @Nested
    @DisplayName("Update Role Tests")
    class UpdateRoleTests {
        
        @Test
        @DisplayName("Should update role successfully")
        void shouldUpdateRoleSuccessfully() {
            // Given
            Long roleId = 1L;
            UpdateRoleRequest request = new UpdateRoleRequest("UPDATED_ADMIN");
            RoleResponse response = new RoleResponse(roleId, "UPDATED_ADMIN", LocalDateTime.now());
            
            when(roleHandler.updateRole(anyLong(), any(UpdateRoleRequest.class)))
                .thenReturn(ResponseEntity.ok(response));
            
            // When
            ResponseEntity<RoleResponse> result = roleController.updateRole(roleId, request);
            
            // Then
            assertEquals(HttpStatus.OK, result.getStatusCode());
            assertNotNull(result.getBody());
            assertEquals(roleId, result.getBody().getId());
            assertEquals("UPDATED_ADMIN", result.getBody().getName());
        }
        
        @Test
        @DisplayName("Should return not found when role does not exist")
        void shouldReturnNotFoundWhenRoleDoesNotExist() {
            // Given
            Long roleId = 999L;
            UpdateRoleRequest request = new UpdateRoleRequest("NEW_NAME");
            
            when(roleHandler.updateRole(anyLong(), any(UpdateRoleRequest.class)))
                .thenReturn(ResponseEntity.notFound().build());
            
            // When
            ResponseEntity<RoleResponse> result = roleController.updateRole(roleId, request);
            
            // Then
            assertEquals(HttpStatus.NOT_FOUND, result.getStatusCode());
            assertNull(result.getBody());
        }
    }
    
    @Nested
    @DisplayName("Delete Role Tests")
    class DeleteRoleTests {
        
        @Test
        @DisplayName("Should delete role successfully")
        void shouldDeleteRoleSuccessfully() {
            // Given
            Long roleId = 1L;
            
            when(roleHandler.deleteRole(anyLong()))
                .thenReturn(ResponseEntity.noContent().build());
            
            // When
            ResponseEntity<Void> result = roleController.deleteRole(roleId);
            
            // Then
            assertEquals(HttpStatus.NO_CONTENT, result.getStatusCode());
            assertNull(result.getBody());
        }
        
        @Test
        @DisplayName("Should return not found when role does not exist")
        void shouldReturnNotFoundWhenRoleDoesNotExist() {
            // Given
            Long roleId = 999L;
            
            when(roleHandler.deleteRole(anyLong()))
                .thenReturn(ResponseEntity.notFound().build());
            
            // When
            ResponseEntity<Void> result = roleController.deleteRole(roleId);
            
            // Then
            assertEquals(HttpStatus.NOT_FOUND, result.getStatusCode());
        }
    }
    
    @Nested
    @DisplayName("Get Role Tests")
    class GetRoleTests {
        
        @Test
        @DisplayName("Should get role by id successfully")
        void shouldGetRoleByIdSuccessfully() {
            // Given
            Long roleId = 1L;
            RoleResponse response = new RoleResponse(roleId, "ADMIN", LocalDateTime.now());
            
            when(roleHandler.getRoleById(anyLong()))
                .thenReturn(ResponseEntity.ok(response));
            
            // When
            ResponseEntity<RoleResponse> result = roleController.getRoleById(roleId);
            
            // Then
            assertEquals(HttpStatus.OK, result.getStatusCode());
            assertNotNull(result.getBody());
            assertEquals(roleId, result.getBody().getId());
            assertEquals("ADMIN", result.getBody().getName());
        }
        
        @Test
        @DisplayName("Should return not found when role does not exist")
        void shouldReturnNotFoundWhenRoleDoesNotExist() {
            // Given
            Long roleId = 999L;
            
            when(roleHandler.getRoleById(anyLong()))
                .thenReturn(ResponseEntity.notFound().build());
            
            // When
            ResponseEntity<RoleResponse> result = roleController.getRoleById(roleId);
            
            // Then
            assertEquals(HttpStatus.NOT_FOUND, result.getStatusCode());
            assertNull(result.getBody());
        }
        
        @Test
        @DisplayName("Should get all roles successfully")
        void shouldGetAllRolesSuccessfully() {
            // Given
            List<RoleResponse> responses = Arrays.asList(
                new RoleResponse(1L, "ADMIN", LocalDateTime.now()),
                new RoleResponse(2L, "USER", LocalDateTime.now())
            );
            
            when(roleHandler.getAllRoles())
                .thenReturn(ResponseEntity.ok(responses));
            
            // When
            ResponseEntity<List<RoleResponse>> result = roleController.getAllRoles();
            
            // Then
            assertEquals(HttpStatus.OK, result.getStatusCode());
            assertNotNull(result.getBody());
            assertEquals(2, result.getBody().size());
            assertEquals("ADMIN", result.getBody().get(0).getName());
            assertEquals("USER", result.getBody().get(1).getName());
        }
        
        @Test
        @DisplayName("Should search roles successfully")
        void shouldSearchRolesSuccessfully() {
            // Given
            String searchPattern = "ADMIN";
            List<RoleResponse> responses = Arrays.asList(
                new RoleResponse(1L, "ADMIN", LocalDateTime.now()),
                new RoleResponse(2L, "SUPER_ADMIN", LocalDateTime.now())
            );
            
            when(roleHandler.searchRoles(anyString()))
                .thenReturn(ResponseEntity.ok(responses));
            
            // When
            ResponseEntity<List<RoleResponse>> result = roleController.searchRoles(searchPattern);
            
            // Then
            assertEquals(HttpStatus.OK, result.getStatusCode());
            assertNotNull(result.getBody());
            assertEquals(2, result.getBody().size());
            assertEquals("ADMIN", result.getBody().get(0).getName());
            assertEquals("SUPER_ADMIN", result.getBody().get(1).getName());
        }
        
        @Test
        @DisplayName("Should get role count successfully")
        void shouldGetRoleCountSuccessfully() {
            // Given
            Map<String, Long> countResponse = Map.of("count", 5L);
            
            when(roleHandler.getRoleCount())
                .thenReturn(ResponseEntity.ok(countResponse));
            
            // When
            ResponseEntity<Map<String, Long>> result = roleController.getRoleCount();
            
            // Then
            assertEquals(HttpStatus.OK, result.getStatusCode());
            assertNotNull(result.getBody());
            assertEquals(5L, result.getBody().get("count"));
        }
    }
}

File: src/test/kotlin/co/onmind/hex/infrastructure/controller/RoleControllerTest.kt
package co.onmind.hex.infrastructure.controller

import co.onmind.hex.application.dto.CreateRoleRequest
import co.onmind.hex.application.dto.RoleResponse
import co.onmind.hex.application.dto.UpdateRoleRequest
import co.onmind.hex.application.handler.RoleHandler
import io.micronaut.http.HttpResponse
import org.junit.jupiter.api.Test
import org.mockito.kotlin.*
import org.assertj.core.api.Assertions.assertThat
import java.time.LocalDateTime

class RoleControllerTest {
    
    private val roleHandler = mock<RoleHandler>()
    private val roleController = RoleController(roleHandler)
    
    @Test
    fun `should delegate create role to handler`() {
        // Given
        val request = CreateRoleRequest("TEST_ROLE")
        val response = RoleResponse(1L, "TEST_ROLE", LocalDateTime.now(), false)
        val httpResponse = HttpResponse.created(response)
        
        whenever(roleHandler.createRole(request)).thenReturn(httpResponse)
        
        // When
        val result = roleController.createRole(request)
        
        // Then
        assertThat(result).isEqualTo(httpResponse)
        verify(roleHandler).createRole(request)
    }
    
    @Test
    fun `should delegate update role to handler`() {
        // Given
        val request = UpdateRoleRequest("UPDATED_ROLE")
        val response = RoleResponse(1L, "UPDATED_ROLE", LocalDateTime.now(), false)
        val httpResponse = HttpResponse.ok(response)
        
        whenever(roleHandler.updateRole(1L, request)).thenReturn(httpResponse)
        
        // When
        val result = roleController.updateRole(1L, request)
        
        // Then
        assertThat(result).isEqualTo(httpResponse)
        verify(roleHandler).updateRole(1L, request)
    }
    
    @Test
    fun `should delegate delete role to handler`() {
        // Given
        val httpResponse = HttpResponse.noContent<Void>()
        whenever(roleHandler.deleteRole(1L)).thenReturn(httpResponse)
        
        // When
        val result = roleController.deleteRole(1L)
        
        // Then
        assertThat(result).isEqualTo(httpResponse)
        verify(roleHandler).deleteRole(1L)
    }
    
    @Test
    fun `should delegate get role by id to handler`() {
        // Given
        val response = RoleResponse(1L, "TEST_ROLE", LocalDateTime.now(), false)
        val httpResponse = HttpResponse.ok(response)
        
        whenever(roleHandler.getRoleById(1L)).thenReturn(httpResponse)
        
        // When
        val result = roleController.getRoleById(1L)
        
        // Then
        assertThat(result).isEqualTo(httpResponse)
        verify(roleHandler).getRoleById(1L)
    }
    
    @Test
    fun `should delegate get all roles to handler`() {
        // Given
        val responses = listOf(RoleResponse(1L, "ADMIN", LocalDateTime.now(), true))
        val httpResponse = HttpResponse.ok(responses)
        
        whenever(roleHandler.getAllRoles()).thenReturn(httpResponse)
        
        // When
        val result = roleController.getAllRoles()
        
        // Then
        assertThat(result).isEqualTo(httpResponse)
        verify(roleHandler).getAllRoles()
    }
    
    @Test
    fun `should delegate search roles to handler`() {
        // Given
        val responses = listOf(RoleResponse(1L, "ADMIN", LocalDateTime.now(), true))
        val httpResponse = HttpResponse.ok(responses)
        
        whenever(roleHandler.searchRoles("ADMIN")).thenReturn(httpResponse)
        
        // When
        val result = roleController.searchRoles("ADMIN")
        
        // Then
        assertThat(result).isEqualTo(httpResponse)
        verify(roleHandler).searchRoles("ADMIN")
    }
    
    @Test
    fun `should delegate get role count to handler`() {
        // Given
        val countMap = mapOf("count" to 5L)
        val httpResponse = HttpResponse.ok(countMap)
        
        whenever(roleHandler.getRoleCount()).thenReturn(httpResponse)
        
        // When
        val result = roleController.getRoleCount()
        
        // Then
        assertThat(result).isEqualTo(httpResponse)
        verify(roleHandler).getRoleCount()
    }
}